esmvalcore/local.py

```diff
@@ -547,11 +494,131 @@ def get_multiproduct_filename(attributes, preproc_dir):
     filename_segments.append(
         f"{attributes['timerange'].replace('/', '-')}.nc")
 
-    outfile = os.path.join(
+    outfile = Path(
         preproc_dir,
         attributes['diagnostic'],
         attributes['variable_group'],
         '_'.join(filename_segments),
     )
 
     return outfile
+
+
+def _path2facets(path: Path, drs: str) -> dict[str, str]:
+    """Extract facets from a path using a DRS like '{facet1}/{facet2}'."""
+    keys = []
+    for key in re.findall(r"{(.*?)}", drs):
+        key = key.split('.')[0]  # Remove trailing .lower and .upper
+        keys.append(key)
+    start, end = -len(keys) - 1, -1
+    values = path.parts[start:end]
+    facets = {key: values[idx] for idx, key in enumerate(keys)}
+    return facets
+
+
+def _select_latest_version(files: list['LocalFile']) -> list['LocalFile']:
+    """Select only the latest version of files."""
+
+    def filename(file):
+        return file.name
+
+    def version(file):
+        return file.facets.get('version', '')
+
+    result = []
+    for _, group in itertools.groupby(sorted(files, key=filename),
+                                      key=filename):
+        duplicates = sorted(group, key=version)
+        latest = duplicates[-1]
```

Comment:
Very good point, I thought I took care of that but it looks like I forgot.

Reply:
There is a test for this case and surprisingly it is passing (which is why I thought I took care of it). Will investigate in more detail..

Reply:
The test passed because for all ESGF data versions start with `v` which automatically comes after `latest`. I changed the code in [787c748](https://github.com/ESMValGroup/ESMValCore/pull/1835/commits/787c748b96a97820afbce2d9ac3f12bed2f57077) so files with a version called `'latest'` are ignored, does it look better now?

Reply:
I think this is fairly safe, calling a version 'latest' is asking for trouble because what will you call the next version? `really_latest_truly_final_v2`? :laughing:

Reply:
I agree that this wouldn't be the best idea, but I've seen a lot of things used in practice that aren't good ideas.

The old version of this code selects `latest` if possible, so why don't you we just stick to that behavior? You claimed that the "changes to the actual file-finding code in this pull request is pretty small"; changing this behavior does not really align with that. I'm fairly sure there is machine/recipe combination out there that relies on that "feature" and will fail or give different results after this change.

Reply:
Here is an example of how the `latest` symlink is used:
```
bandela@sci4:~$ ll /badc/cmip6/data/CMIP6/CMIP/BCC/BCC-ESM1/historical/r1i1p1f1/Amon/tas/gn/
total 1
drwxr-x--- 1 badc open    0 13 mrt  2019 files
lrwxrwxrwx 1 badc badcint 9 13 mrt  2019 latest -> v20181214
drwxr-x--- 1 badc open    0 13 mrt  2019 v20181214
```

> The old version of this code selects latest if possible, so why don't we just stick to that behavior?

Because it's impossible to know what version the data hiding behind `latest` is. So with this behaviour, we cannot record the version and therefore we cannot have reproducible recipes because the next time you run a recipe the data may have changed. If you think this is a really important feature, we could make it configurable from config-user.yml?

Reply:
Or I could make it smarter, so files with version `'latest'` are only removed from the search result if they are the same as another file. Would that be better?

Reply:
> Or I could make it smarter, so files with version `'latest'` are only removed from the search result if they are the same as another file. Would that be better?

Yes, I think that's a good idea! Thanks!

Reply:
Done in [098ca88](https://github.com/ESMValGroup/ESMValCore/pull/1835/commits/098ca88f4c0f4c4494d23db25369c40ae6b690fd)